tests/unit/test.config.js

```diff
@@ -290,34 +289,86 @@ describe('config', () => {
         'source-dir': 'fake/value/',
       };
       assert.throws(() => {
-        applyConf({argv, configObject, defaultValues});
+        applyConf({...params, configObject});
       }, UsageError, 'UsageError: The config option "source-dir" must be ' +
         'specified in camel case: "sourceDir"');
     });
 
     it('throws an error when an option is invalid', () => {
-      const {argv, defaultValues} = makeArgv({
+      const params = makeArgv({
         globalOpt: {
           'source-dir': {
             type: 'string',
             demand: false,
           },
         },
       });
-      const configFileName = 'fake/path/to/config';
       const configObject = {
-        artifactsDir: 'fake/artifacts/dir',
+        randomDir: 'fake/artifacts/dir',
       };
       assert.throws(() => {
-        applyConf({argv, configObject, defaultValues, configFileName});
-      }, UsageError, 'UsageError: The config file at fake/path/to/config ' +
-        'specified an unknown option: "artifactsDir"');
+        applyConf({...params, configObject});
+      }, UsageError, 'UsageError: The config file ' +
+      'at some/path/to/config.js specified an unknown option: "randomDir"');
+    });
+
+    it('throws an error when a global option type is invalid', () => {
+      const params = makeArgv({
+        globalOpt: {
+          retries: {
+            type: 'number',
+            default: 1,
+          },
+        },
+      });
+      const configObject = {
+        retries: 'invalid-value',
+      };
+      assert.throws(
+        () => applyConf({...params, configObject}),
+        UsageError,
+        'UsageError: The config file at some/path/to/config.js specified the ' +
+        'type of "retries" incorrectly as "string" (expected type: "number")'
+      );
+    });
+
+    it('throws an error when the type of option value is invalid', () => {
+      const params = makeArgv({
+        globalOpt: {
+          'source-dir': {
+            type: 'string',
+            demand: false,
+          },
+        },
+      });
+      const configObject = {
+        sourceDir: {randomKey: 'randomValue'},
+      };
+      assert.throws(() => {
+        applyConf({...params, configObject});
+      }, UsageError, 'UsageError: The config file at some/path/to/config.js ' +
+        'specified the type of "sourceDir" incorrectly');
+    });
+
+    it('does not throw an error when the type of option value is count', () => {
+      const params = makeArgv({
+        globalOpt: {
+          verbose: {
+            type: 'count',
+          },
+        },
+      });
+      const configObject = {
+        verbose: 2,
+      };
+      const newArgv = applyConf({...params, configObject});
+      assert.strictEqual(newArgv.verbose, 0);
```

Comment:
I dug into this. This test works fine if a default value is explicitly specified in the option definition, but otherwise it triggers the `wasValueSetOnCli` condition to be true, assigning a default value of 0 (because the type is `count`, I guess) & ignores the configured value.